core/src/family.cc

```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
 
 template <typename T>
 std::vector<MetricFamily> Family<T>::Collect() {
-  std::lock_guard<std::mutex> lock{mutex_};
   auto family = MetricFamily{};
-  family.name = name_;
-  family.help = help_;
-  family.type = T::metric_type;
-  for (const auto& m : metrics_) {
-    family.metric.push_back(std::move(CollectMetric(m.first, m.second.get())));
+  std::vector<std::shared_ptr<T>> to_be_removed;
+  {
+    std::lock_guard<std::mutex> lock{mutex_};
+    family.name = name_;
+    family.help = help_;
+    family.type = T::metric_type;
+    for (const auto& m : metrics_) {
+      if (!m.second->Expired()) {
+        family.metric.push_back(std::move(CollectMetric(m.first, m.second)));
+      } else if (retention_behavior_ == RetentionBehavior::Remove) {
+        to_be_removed.push_back(m.second);
+      }
+    }
+  } // end mutex
+  for (auto metric: to_be_removed) {
+    Remove(metric);
+  }
+  if (family.metric.empty()) {
+    return {};
+  } else {
+    return {family};
+  }
+}
+
+template <typename T>
+bool Family<T>::UpdateRetentionTime(const double& retention_time, const std::string& re_name, 
```

Comment:
Couldn't you just update the retention time for each metric internally every time the metric is updated (incremented, decremented etc.) and set a flag for removal every time the retention time is exceeded to circumvent the name/metric matching process? 

Even if it remains as is, shouldn't this be a method that's called internally by the registry? Should the user explicitly have to call this function if they're set the retention behavior for a metric to remove?

Reply:
The rentention time is updated automatically, these are just for the developer to keep the metric alive if they are not updating the metric with new data. As for flagging them for removal, we want to flag them peroidically, otherwise we might get a metric that gets data once during the whole program and then never expires.

The current implementation has the metrics update the class member last_update_ instead of calling UpdateRetentionTime since that method is used to find the metric or group of metrics and update their retention time.